text/0486-deprecate-mouseenter.md

```diff
@@ -0,0 +1,162 @@
+the corresponding event handler methods (`mouseEnter() {}`) in Ember components and 
+`{{action "some" on="mouseenter"}}`. 
+
+## Motivation
+
+Ember's EventDispatcher handles "Ember events" by attaching listeners to the app's root element
+and relying on the events bubbling up to that element (aka event delegation). There they 
+are processed and invoke any matching `Ember.Component` event handler method with 
+the same (camel-cased) method as the event type. Same for element-space `{{action}}` 
+modifiers.
+
+> Note: for a "Deep Dive on Ember Events" and how they differ from "native events" I refer to 
+Marie Chatfield's excellent 
+[blog post](https://medium.com/square-corner-blog/deep-dive-on-ember-events-cf684fd3b808)
+or [EmberConf talk](https://youtu.be/G9hXjjHFJVs)
+
+This works fine in general, but `mouseenter`/`mouseleave` events are special as they do
+not bubble. In the past it still worked nevertheless as jQuery transparently handled this
+for us, as it had special support for event delegation for these events, essentially by using 
+(bubbling) `mouseover` events to replicate the semantics of `mouseenter`/`mouseleave` events.
+
+When support for jQuery-less apps was introduced, this [left a hole](https://github.com/emberjs/ember.js/issues/16591)
+in the jQuery-less EventDispatcher implementation. But as support for those events was and
+still is part of Ember's pubic API, we had no chance other than to [implement support
+for jQuery-less apps](https://github.com/emberjs/ember.js/pull/16603) using the same 
+`mouseover` based approach.
+
+This however comes with a cost: besides an [unresolved issue](https://github.com/emberjs/ember.js/issues/17228)
+the implementation has some performance drawbacks, as it has to process every `mouseover` event on 
+*any* element, create fake `mouseenter`/`mouseleave` events and try to dispatch them, even when 
+not a single component/action needs them.  
+
+Deprecating support for `mousemove` is also proposed, which is a (bubbling) event that does not have the higher 
+implementation cost as `mouseenter`/`mouseleave`, but nevertheless requires the EventDispatcher to optimistically handle
+these extremely high-volume events.
+
+While efforts to make this more "pay as you go" are [possible](https://github.com/emberjs/ember.js/pull/17911), 
+the trade-off of keeping support around still seems unfavorable, as these events fire so
+frequently, while they are (most certainly) very rarely used.
+
+This is even more so given that Glimmer Components with their outerHTML semantics do not 
+work with event handler methods, and `{{action}}` will eventually fade away in favor of
+`{{on}}` using native `addListener()`.
+
+## Transition Path
+
+When a `mouseenter`/`mouseleave`/`mousemove` event is dispatched by the EventDispatcher to a component
+or `{{action}}` listening to that event type, a deprecation warning will be issued.
```

Comment:
I already started implementing this, which made me think about this again. I stated that the deprecation is triggered when the event is dispatched. But this can be problematic, as it relies on a specific user interaction. So there might be cases where this deprecation fires *very* frequently, while in others it might fire very *rarely* (so the developer does not even see it). 

Maybe it's better to trigger that at *creation* time (i.e. in the `CurlyComponentManager` / `ActionModifierManager`). Thoughts?

Reply:
Yes, indeed. We should fire on component creation (_basically_ the same way we assert when you use event methods in tagless Ember.Component), and for {{action}} we’d do it at template compilation (when it’s a string) or initial render (when the event name is dynamic).